library/alloc/src/string.rs

```diff
@@ -1628,11 +1628,13 @@ impl String {
         self.vec.clear()
     }
 
-    /// Creates a draining iterator that removes the specified range in the `String`
-    /// and yields the removed `chars`.
+    /// Removes the specified range from the string, returning all removed
+    /// characters as an iterator.
     ///
-    /// Note: The element range is removed even if the iterator is not
-    /// consumed until the end.
+    /// # Leaking
+    ///
+    /// In case the iterator disappears without getting dropped (using
+    /// [`core::mem::forget`], for example), the range remains in the string.
```

Comment:
I don’t understand why this would need to be specified. It encourages usage of `mem::forget` which usually shouldn’t be used on foreign types.

Reply:
See my description.

Reply:
Or clarify what you mean with "this": the entire paragraph, or the constraint about what can happen if you leak an iterator?

Reply:
Well, this paragraph kind-of says to me "If you want to, during iteration, change your mind and keep the range inside of the original string, feel free to go ahead and just `mem::forget` the iterator. This is a supported use-case and we guarantee the precise behavior in this case". I.e. it promises a certain behavior on `mem::forget`ting the iterator, and thus (implicitly) ***encourages*** people to deliberately leak the iterator to achieve certain behavior. If we *wanted* to offer an actual documented and guaranteed way of avoiding the draining, then it should be via a (`self`-consuming) method on the iterator, not by giving any promises on "what happens on `mem::forget` IMO.  (E.g. one could create a `fn cancel(self)` on the `string::Drain` iterator that explicitly doesn't remove anything from the `String`, or `vec::Drain` could provide a method that puts back the remaining elements into the original `Vec`.)

All we should do in these kinds of documentation is say, very explicitly, something along the lines of "please ***don't ever*** skip dropping these iterators properly, it is considered a logic error and can, among other things, leave the original structure in an arbitrary unspecified (yet safe-to-use in terms of memory-safety) state with potentially surprising and unexpected behavior when you continue using it afterwards, it can result in arbitrary amounts of additional memory leaks of the drained items and/or other items in the same data structure". I.e. ***strongly discourage*** actually leaking the iterators, and perhaps provide a non-exhaustive list of things that *can* happen upon leaking it anyways, without giving any promises that constrain the set of things that can happen any more than necessary by Rust's memory safety guarantees.

Reply:
>See my description.

Do you mean this sentence?

> Apparently people do use `mem::forget` a lot and are surprised by its effect, see the history of the notice on `vec::drain`.

If "the history of the notice on `vec::drain`" is supposed to explain this, can you perhaps provide some links for me to read up on this?

Reply:
> > See my description.
> 
> Do you mean this following sentence?

I mean primarily that the description already lists reasons why it would need to be specified as well as reasons not to.

> > Apparently people do use `mem::forget` a lot and are surprised by its effect, see the history of the notice on `vec::drain`.
> 
> If "the history of the notice on `vec::drain`" is supposed to explain this, can you perhaps provide some links for me to read up on this?

I went over #43244 again and don't see it. It's in the existence of #74652 and its predecessors introducing the paragraph. But I probably misinterperted #73844: it's rhe description that inspired the clarification, not someone actually using mem::forget and expecting to still look at the vector later. Some other posts too I'm sure, but probably less then I thought.

Reply:
> `vec::Drain` could provide a method that puts back the remaining elements into the original `Vec`.)

Intriguing idea, I think no one mentioned that possibility yet.
 
> All we should do in these kinds of documentation is say, very explicitly, something along the lines of

I agree, but apparently no one wanted to discourage users in the Vec::drain description.